pkg/endpoint/providers/kube/client.go

```diff
@@ -210,20 +208,25 @@ func (c *Client) run(stop <-chan struct{}) error {
 
 // getServicesByLabels gets Kubernetes services whose selectors match the given labels
 func (c *Client) getServicesByLabels(matchLabels map[string]string, namespace string) ([]corev1.Service, error) {
-	var serviceList []corev1.Service
-	svcList, err := c.kubeClient.CoreV1().Services(namespace).List(context.Background(), metav1.ListOptions{})
-	if err != nil {
-		log.Error().Err(err).Msgf("Error listing Services in namespace %s", namespace)
-		return nil, err
+	var namespacedServiceList []*corev1.Service
+	var finalList []corev1.Service
+
+	serviceList := c.kubeController.ListServices()
+	// TODO: #1684 Introduce APIs to dynamically allow applying selectors, instead of callers implementing
+	// filtering themselves
+	for _, svc := range serviceList {
+		if svc.Namespace == namespace {
+			namespacedServiceList = append(namespacedServiceList, svc)
+		}
 	}
 
-	for _, svc := range svcList.Items {
+	for _, svc := range namespacedServiceList {
 		svcRawSelector := svc.Spec.Selector
 		selector := labels.Set(svcRawSelector).AsSelector()
 		if selector.Matches(labels.Set(matchLabels)) {
-			serviceList = append(serviceList, svc)
+			finalList = append(finalList, *svc)
```

Comment:
Why can't this loop be merged with the above loop when the services are already being walked through?

Reply:
It actually could, I missed that